vermouth/gmx/itp_read.py

```diff
@@ -0,0 +1,533 @@
+
+        Raises
+        ------
+        KeyError
+            If the section header is unknown.
+        """
+                        
+        prev_section = self.section
+
+        ended = []
+        section = self.section + [line.strip('[ ]').casefold()]
+        if tuple(section[-1:]) in self.METH_DICT:
+            self.section = section[-1:]
+        else:
+            while (tuple(section) not in self.METH_DICT
+                   and len(section) > 1):
+                ended.append(section.pop(-2))  # [a, b, c, d] -> [a, b, d]
+            self.section = section
+                          
+        result = None
+
+        if len(prev_section) != 0:
+            print(len(prev_section))
+            result = self.finalize_section(prev_section, ended)
+
+        action = self.header_actions.get(tuple(self.section))
+        if action:
+            action()
+                           
+        return result
+
+    def finalize_section(self, pervious_section, ended_section):
+
+        """
+        Called once a section is finished. It appends the current_links list
+        to the links and update the block dictionary with current_block. Thereby it
+        finishes the reading a given section. 
+
+        Parameters
+        ---------
+        previous_section: list[str]
+            The last parsed section.
+        ended_section: list[str]
+            The sections that have been ended.
+        """
+        
+        if self.current_block is not None:
+           self.force_field.blocks[self.current_block.name] = self.current_block
+       
+    def _new_block(self):
+        self.current_block = Block(force_field=self.force_field)
+
+    @SectionLineParser.section_parser('moleculetype')
+    def _block(self, line, lineno=0):
+        name, nrexcl = line.split()
+        self.current_block.name = name
+        self.current_block.nrexcl = int(nrexcl)
+
+    @SectionLineParser.section_parser('moleculetype', 'atoms')
+    def _block_atoms(self, line, lineno=0):
+        tokens = collections.deque(_tokenize(line))
+        _parse_block_atom(tokens, self.current_block)
+
+    @SectionLineParser.section_parser('moleculetype', 'bonds', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'angles', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'dihedrals', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'impropers', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'constraints', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'pairs', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'exclusions', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'virtual_sites2', context_type='block')
+    @SectionLineParser.section_parser('moleculetype', 'position_restraints', context_type='block')
+
+    def _interactions(self, line, lineno=0, context_type=''):
```

Comment:
This code will only ever process blocks, right? So factor out the context_type argument.

Also, shouldn't this create molecules?

Reply:
I think we did not conclusively discuss if ITPs should spawn molecules or blocks. This also raises the questions if then the ITPdirector should populate a system instead of a force-field. I think it is OK to consider an ITP to be a block. This would sort of keep it streamline with Martinize idea to build larger molecules from blocks. It has the disadvantage of the debate of index versus atomname for nodes. I can already tell you all my tests I did for the polymers tell me that both are exactly equivalent, meaning as long as links and blocks are both either name or index based the entire code just works, which is great news on its own.